native-dnn/src/main/java/com/intel/analytics/bigdl/mkl/hardware/platform/linux/LinuxAffinity.java

```diff
@@ -2,84 +2,185 @@
 
 import com.intel.analytics.bigdl.mkl.hardware.platform.IAAffinity;
 
-import java.util.Arrays;
-import java.util.BitSet;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class LinuxAffinity implements IAAffinity {
-    public static native int setAffinity0(int[] sets);
-    public static native int getAffinity0(int[] sets, int cpuCounts);
-    public static native void setOmpAffinity0(int[] sets);
+    public static native int setAffinity0(int[] set);
+    public static native int getAffinity0(int[] set);
+    public static native int[] setOmpAffinity0(int[] set);
+    public static native int[] getOmpAffinity0(int cpuCounts);
 
     private int cpuCounts;
-    private boolean[] availableCPUs;
+    private int[] cpuList;
+    private ConcurrentHashMap<Integer, List<Long>> availableCores;
 
     public LinuxAffinity(int cpuCounts) {
         this.cpuCounts = cpuCounts;
-        this.availableCPUs = new boolean[cpuCounts];
-        Arrays.fill(this.availableCPUs, false);
+
+        this.cpuList = getAffinity();
+
+        this.availableCores = new ConcurrentHashMap<Integer, List<Long>>();
+        for (int key : this.cpuList) {
+            availableCores.put(key, new ArrayList<Long>());
+        }
     }
 
-    public synchronized int setAffinity() {
-        int ret = -1;
-        for (int i = 0; i < availableCPUs.length; i++) {
-            if (!availableCPUs[i]) {
-                ret = i;
-                int[] set = {i};
-                setAffinity0(set);
-                availableCPUs[i] = true;
-                break;
+    public synchronized void setAffinity() {
+        int min = Integer.MAX_VALUE;
+        int targetCore = 0; // by default, use the first core of cpuList.
+        for (int key : this.cpuList) {
+            int size = availableCores.get(key).size();
+            if (size < min) {
```

Comment:
so here you want to select core which has least binding threads ? why not use `TreeHashMap` then

Reply:
Yes. Do you mean `TreeMap`? It's not thread safe.